Add vertical alignment to diagram renderer

zickgraf opened the pull request:
This is optional and off by default.

I have not implemented an options object for `render()` as suggested in  https://github.com/yishn/tikzcd-editor/issues/27#issuecomment-425070268 since I was not sure how to deal with the existing `co` option in this case.

yishn commented:
Thanks for the pull request! We can leave `co` as it is, e.g. as a separate argument. It's not documented and should only be used internally. I'd like the `align` property be realized as a key in an `options` object, as the second argument to `render` and `corender`.

yishn commented on src/components.js:
```diff
@@ -125,23 +125,34 @@ export class Diagram extends Component {
+            ].join(' ')
+        ))
+
+        if(this.props.align && cells.length > 0) {
```
To make the code style consistent, please leave a space after `if` and `for`, right before `(`.

yishn commented on src/components.js:
```diff
@@ -125,23 +125,34 @@ export class Diagram extends Component {
+        ))
+
+        if(this.props.align && cells.length > 0) {
+            for(let j in cells[1]) {
```
A `for` `in` loop over arrays is considered [bad practice](https://stackoverflow.com/questions/500504/why-is-using-for-in-with-array-iteration-a-bad-idea). Please use `arr.forEach()` or a normal `for` loop.

yishn commented on src/components.js:
```diff
@@ -125,23 +125,34 @@ export class Diagram extends Component {
+
+        if(this.props.align && cells.length > 0) {
+            for(let j in cells[1]) {
+                let maxLength = Math.max.apply(null, cells.map(entries => entries[j].length));
```
Please remove all semicolons for code style consistency.

yishn replied:
Instead of `Math.max.apply(null, arr)` you can use the ES6 syntax `Math.max(...arr)`. It's more readable IMHO.

yishn commented on src/components.js:
```diff
@@ -125,23 +125,34 @@ export class Diagram extends Component {
+            for(let j in cells[1]) {
+                let maxLength = Math.max.apply(null, cells.map(entries => entries[j].length));
+                for(let i in cells) {
+                    cells[i][j] = cells[i][j].padEnd(maxLength, " ");
```
Please use `'` instead of `"` for code style consistency.

yishn commented on src/render.js:
```diff
@@ -19,7 +19,7 @@ function resolveComponents(vnode) {
 }
 
-export function renderToDiagram(vnode, co = false) {
+export function renderToDiagram(vnode, align = false, co = false) {
```
Please use an `options` object as discussed in https://github.com/yishn/tikzcd-editor/issues/27#issuecomment-425070268.

yishn commented on src/render.js:
```diff
@@ -28,14 +28,15 @@ export function renderToDiagram(vnode, co = false) {
     return new Diagram({
         ...diagramNode.props,
         co: co !== !!diagramNode.props.co,
+        align: align,
```
In ES6, you don't need to write `align: align,` in an object literal, simply `align,` will suffice.

zickgraf commented:
I've addressed the issues (and learned some new ES6 features I really like :D ). Regarding the `options` object: I am not sure if I should pass it to the `render` function of the diagram via a property of the diagram or as an argument. For now, I have chosen the first approach.

yishn commented:
Ah, I don't think we can pass `options` directly as a prop to `Diagram`. We already use the `options` prop on `Diagram` for [tikzcd environment options](https://github.com/yishn/jsx-tikzcd/pull/3/files#diff-3448e707cf244e61238798459dcae24cR153). Hm... maybe we shouldn't pass the `options` object directly to `Diagram` [here](https://github.com/yishn/jsx-tikzcd/pull/3/files#diff-855b05ea6fe92e5b91d6da40f15dae9fR31) but rather only the `align` property.

So, the `(co)render()` functions in `render.js` receive an `options` object as second function argument, but the `Diagram` class receives only the `align` prop.

zickgraf commented:
Okay, now `options` is unpacked in `renderToDiagram` with an explicit fallback to `false` (instead of the implicit fallback to `undefined`).

yishn commented:
I believe that's correct now, thank you so much!
